Source/Dafny/Dafny.atg

```diff
@@ -2625,9 +2625,9 @@ Rhs<out AssignmentRhs r>
        }
     .)
   | "*"                                (. r = new HavocRhs(t); .)
-  | Expression<out e, false, true>     (. r = new ExprRhs(e); .)
+  | Expression<out e, false, true>     (. r = e != null ? new ExprRhs(e) : null; .)
```

Comment:
How can the parser for Expression return `null`, isn't that the bug?

Reply:
Interesting idea.

 `Expression<out e, false, true>` will end up calling through the chain `EquivExpression, ImpliesExpliesExpression, LogicalExpression` the method
 `RelationalExpression` which will set e to a dummy expression. So far so good, there should be no crash.
However, down the call chain, it goes through `ShiftTerm`, `Term`, `Factor`, `BitvectorFactor`, `AsExpression`, `UnaryExpression`, `PrimaryExpression`.
At this point, `PrimaryExpression` set `e` to `null` and then tries to call `IsSetDisplay()` to parse the multiset. However, this test actually lookup two tokens which should be `multiset(` or `multiset{`, but since it sees `multiset}`, none of the tests pass and it eventually calls `SynErr(279)`, but `e` is still set to null.

This PR was fixing the surface bug. However, it seems that from your remark and my findings, we might as well remove the setting of `e` to  `null` in `PrimaryExpression` or at least put a dummy value instead, so that we wouldn't run into the same case later for other expressions.

I'm testing this hypothesis by pushing a commit that ensures that PrimaryExpression never returns null.

Reply:
I see. I misunderstood how CocoR handles parse failures. I thought that chaining an empty parse result up the parse stack was handled by CocoR, but it seems we have to do that ourselves by checking for `null` everywhere and then propagating more null upwards.

I'm not sure I like the dummy expression approach, since it hides the fact that we couldn't produce an Expression by nesting the dummyExpr in other nodes like `ExprRhs`.